Module 3: CRUD and data models (#4)
ReCoded-Org/curriculum-backend-readings: module3-crud-and-data-models/r1-schema-and-data-models/README.md

## Intoduction to database schemas

A database schema is a blueprint or architecture of how our data will look. It doesn’t hold data itself, but instead describes the shape of the data and how it might relate to other tables or models.

For example, this is a `customer` schema in MongoDB using mongoose.

```js
const mongoose = require("mongoose");

const customer = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  zipcode: {
    type: Number,
  },
});

module.exports = mongoose.model("Customer", customer);
```

The same `customer` schema in a MySql database.

```sql
CREATE TABLE customer (
 id INT AUTO_INCREMENT PRIMARY KEY,
 name VARCHAR(50) NOT NULL,
 zipcode INT(15) default NULL,
)

```

We can see that the schema definition for the `customer` has a `name` which is a type `string` and a `zipcode` that is of type `number`.

### The Difference between data model and schema

> The database schema is one that contains list of attributes and instructions to tell the database engine how data is organised whereas data model is a collection of conceptional tools for describing data, data-relationship and consistency constraints.

Reviewer: This needs an example to be well understood. Data modeling could be related to multiple objects or resources. While schema is for one object shape in the database.

Author: Agree, it needs an example, however, looking at this now doesn't seem right to be here, I think it's already similar to what is in module 2 as Louis said above, I will just remove it from here.

REST is a set of architectural constraints. API developers can implement REST in a variety of ways. When a client request is made via a RESTful API, it transfers a representation of the state of the resource to the requester or endpoint. This information, or representation, is delivered in one of several formats, JSON is the most generally popular file format to use.

### CRUD (CREATE, READ, UPDATE, DELETE)

In a REST environment, CRUD often corresponds to the HTTP methods `POST`, `GET`, `PUT`, and `DELETE`, respectively. These are the fundamental elements of a persistent storage system.

For example, imagine we are working with a system that is keeping track of meals and their corresponding prices for a restaurant. Let’s look at how we would implement CRUD operations.

#### Create

To create resources in a REST environment, we most commonly use the HTTP POST method. POST creates a new resource of the specified resource type.
Imagine that we are adding a new food item to the stored list of dishes for this restaurant, and the dish objects are stored in a dishes resource.

Note: Use `POST` when you want to add a child resource under resources collection.

##### Request:

> `POST http://www.myrestaurant.com/api/dishes/`

Reviewer: can we add an example using axios or fetch with data?

Author: I didn't add an example as they should know how to do that with JS, the point here is just to emphasize the explanation above, maybe adding a `fetch` can confuse them as we are talking about backend and suddenly inject some frontend in?

Reviewer: I agree. but also the given example here is incomplete missing data, and I can't see what learning would it add

We will need to send the `dish` data too.

#### Read

To read resources in a REST environment, we use the GET method. in practice, reading a resource shouldn't change any information - it should only retrieve it.

To read resources in a REST environment, we use the GET method. Reading a resource should never change any information - it should only retrieve it. REST itself is more like a set of guidelines. Technically, you can change data in a `GET` request, but since we are creating a RESTful API, you shouldn't do that. Having a GET request that updates data in your database would be confusing to the users of your API and violate the expected behavior of REST.

##### Request:

`GET http://www.myrestaurant.com/api/dishes/`

#### Update

PUT is the HTTP method used for the CRUD operation, Update.
So if the price of Avocado Toast has gone up, we should go into the database and update that information using PUT.

Note: Use `PUT` when you want to modify a singular resource which is already a part of resources collection. `PUT` replaces the resource with the data you send in its entirety.

##### Request:

> `PUT http://www.myrestaurant.com/dishes/1223`

Reviewer: emphasize the addition of resource id to the url

#### Delete

The CRUD operation Delete corresponds to the HTTP method DELETE. It is used to remove a resource from the system.

> Let’s say that the world avocado shortage has reached a critical point, and we can no longer afford to serve this modern delicacy at all. We should go into the database and delete the item that corresponds to `"Avocado Toast"`, which we know has an `id` of 1223.

Reviewer: `"Avocado Toast"` is over-emphasized. Use _italic_ or **bold** without quotations.

## Introduction To Database Schemas

A database schema is a blueprint or architecture of how our data will look. It doesn’t hold data itself, but instead describes the shape of the data and how it might relate to other tables or models.

For example, this is a `customer` schema in MongoDB using Mongoose.

```js
const mongoose = require("mongoose");

const customer = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  zipcode: {
    type: Number,
  },
});

module.exports = mongoose.model("Customer", customer);
```

> The same `customer` schema in a MySql database.

Reviewer 2: ```suggestion
The same `customer` schema in a MySQL database.
```

```sql
CREATE TABLE customer (
 id INT AUTO_INCREMENT PRIMARY KEY,
 name VARCHAR(50) NOT NULL,
 zipcode INT(15) default NULL,
)

> ```

Reviewer 2: This is the SQL command to create a database.
You can also give an example to show a schema in PostgreSQL using Sequelize, similar to how you've shown MongoDB schema using Mongoose. So that students are not confused between schema using an ORM and SQL commands.

We can see that the schema definition for the `customer` has a `name` that is type `string` and a `zipcode` that is of type `number`.

## Introduction to CRUD operations

When we are building APIs, we want our models to provide four basic types of functionality. The model must be able to Create, Read, Update, and Delete resources. Computer scientists often refer to these functions by the acronym CRUD. A model should have the ability to perform at most these four functions in order to be complete.

### CRUD and REST

**What is REST?**

REST is a set of architectural constraints. API developers can implement REST in a variety of ways. When a client request is made via a RESTful API, it transfers data as an object in some format to the requester or endpoint. There are several formats, JSON is the most generally popular file format to use.

**Resource in REST**

> A resource in REST is a like an Object or an Entity in a Database. Once a resource is identified then its representation is to be decided using a standard format so that the server can send the resource in the decided on format and client can understand the same format. A resource can be any information that can be named (e.g. a person, a user, an invoice, a collection of invoices, etc).

Reviewer 2: REST is explained in detail in module 1, so this part can be removed.
You could instead rename this section as CRUD in REST, and explain with examples that when handling different REST operations like GET, PUT, POST, DELETE there is most likely an underlying database CRUD operation taking place. Otherwise the examples being shared after this will also be redundant with module 1.

Author: After going through module 1, this makes sense, however, I think this process would have been better if more details were provided in the module outline sheet, like _"explain with examples that when handling different REST operations like GET, PUT, POST, DELETE there is most likely an underlying database CRUD operation taking place"_, so I would know easily what to write and explain here.

## Intoduction to schemas and data modeling

### Data modeling

> A database model is a type of data model that determines the logical structure of a database. It is the high level design which defines the kind of tables, the `fields` in those tables and the `relations` between different tables.

Reviewer 3: you shouldnt be using `code` for non-code terms, it's not really a thing. bold it if it's a term. and also, if you are assuming they know what fields and relations are, you dont need to bold them (it's explained in module 2, and also you didn't write any definitions)

> The most popular example of a database model is the relational model, which uses a table-based format, which is the one we use in most SQL databases.

Reviewer 3: redundant with module 2 imo...

### What are database schemas?

> A database schema is a `blueprint` or `architecture` of how our data will look. It doesn’t hold data itself, but instead describes the shape of the data and how it might relate to other tables or models.

Reviewer 3: same comment about using `code` type here

> For example, This is a `Customer` schema in mongoDB using mongoose.

Reviewer 3: MongoDB

```js
const mongoose = require("mongoose");

const customer = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  zipcode: {
    type: Number,
  },
});

module.exports = mongoose.model("Customer", customer);
```

The same `Customer` schema in a MySql database.

```sql
CREATE TABLE customer (
 id INT AUTO_INCREMENT PRIMARY KEY,
 name VARCHAR(50) NOT NULL,
 zipcode INT(15) default NULL,
)

```

We can see that the schema definition for the `Customer` has a `name` which is a type `string` and a `zipcode` that is of type `number`.

> ### The Difference between data model and schema

Reviewer 3: fix the capitalization to be consistent

> The database schema is one that contains list of attributes and instructions to tell the database engine how data is organised whereas Data model is a collection of conceptional tools for describing data, data-relationship and consistency constraints.

Reviewer 3: Data model randomly capitalized

## Introduction to CRUD operations

When we are building APIs, we want our models to provide four basic types of functionality. The model must be able to Create, Read, Update, and Delete resources. Computer scientists often refer to these functions by the acronym `CRUD`. A model should have the ability to perform at most these four functions in order to be complete.

### CRUD and REST

In a REST environment, CRUD often corresponds to the HTTP methods `POST`, `GET`, `PUT`, and `DELETE`, respectively. These are the fundamental elements of a persistent storage system.

For example, imagine we are working with a system that is keeping track of meals and their corresponding prices for a restaurant. Let’s look at how we would implement CRUD operations.

#### Create

To create resources in a REST environment, we most commonly use the HTTP POST method. POST creates a new resource of the specified resource type.
Imagine that we are adding a new food item to the stored list of dishes for this restaurant, and the dish objects are stored in a dishes resource.

> Request:

Reviewer 3: use headers of some sort here, otherwise it doesnt look good

`POST http://www.myrestaurant.com/api/dishes/`

We will need to send the `dish` data too.

Body -

```json
{
  "dish": {
    "name": "Avocado Toast",
    "price": 8
  }
}
```

> Response: Status Code - 201 (Created)

Reviewer 3: are they supposed to know what request response and body are already at this point?

This creates a new item with a `name` value of `“Avocado Toast”` and a price value of 8. Upon successful creation, the server should return a header with a link to the newly-created resource.

#### Read

> To read resources in a REST environment, we use the GET method. Reading a resource should never change any information - it should only retrieve it.

Reviewer 3: you should explain that this is not strictly enforced in code but by convention, and also i think in general you need more explaining that REST is a convention. you can technically do whatever you want but it would be confusing to users of your API

> Request:

Reviewer 3: or use bold, you dont have to use headers, but not "Response:"

Reviewer 3: fix

`GET http://www.myrestaurant.com/api/dishes/`

Response: Status Code - 200 (OK)

Your response should have a list of `dishes`.

Body -

```json
{
  "dishes": [
    {
      "id": 1,
      "name": "Spring Rolls",
      "price": 6
    },
    {
      "id": 2,
      "name": "Mozzarella Sticks",
      "price": 7
    },
    {
      "id": 1223,
      "name": "Avocado Toast",
      "price": 8
    },
    {
      "id": 1224,
      "name": "Muesli and Yogurt",
      "price": 5
    }
  ]
}
```

> #### Update

Reviewer 3: you should explain when to use PUT vs POST

> ## Introduction To Database Schemas

Reviewer 2: Every reading should start with an H1 for the reading title, then it should have the summary and learning goals. Then each main topic is an H2, any further sub-topics or sections within them can be H3, H4 or a bold highlight.

These are there in the contributing guidelines in the repo README. You can also refer module 1 readings.

> A database schema is a blueprint or architecture of how our data will look. It doesn’t hold data itself, but instead describes the shape of the data and how it might relate to other tables or models.

Reviewer 2: ```suggestion
A database schema is a blueprint or architecture of how our data will look. It doesn’t hold data itself, but instead describes the shape and structure of the data and how it might be related to other data.
```

Since we haven't yet introduced the meaning of "models".

For example, this is a `customer` schema in MongoDB using Mongoose.

```js
const mongoose = require("mongoose");

const customer = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  zipcode: {
    type: Number,
  },
});

module.exports = mongoose.model("customer", customer);
```

> The same `customer` schema in PostgreSQL database using sequelize.

Reviewer 2: ```suggestion
The same `customer` schema in PostgreSQL using [Sequelize](https://sequelize.org/).
```
Remove "database" after PostgreSQL, similar to how you are just saying MongoDB above. Always keep S of Sequelize in capital letter, and add the link to https://sequelize.org/.

```js
const Sequelize = require("sequelize");

module.exports = (sequelize, DataTypes) => {
  const Customer = sequelize.define("customer", {
    name: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    zipcode: {
      type: DataTypes.INTEGER,
    },
  });
  return Customer;
};
```

> We can see that the schema definition for the `customer` has a `name` that is type `string` and a `zipcode` that is of type `number`.

Reviewer 2: ```suggestion
We can see that the schema definition for the `customer` has a `name` that is type `string` and a `zipcode` that is of type `number`. The `name` field is mandatory and cannot be empty.
```

> ## Introduction to CRUD operations

Reviewer 2: In the previous section, I think we will need some explanation of what a "model" is. Otherwise we are mentioning this word in this new section without students being properly introduced to its meaning.

To be specific we can explain similar to how [this Mongoose article](https://mongoosejs.com/docs/models.html) does, given that all labs will be in Mongoose for this module.

      // the { new: true } here is just to tell Mongoose we want the newely updated dish back
      new: true,
    });
    // we return the updated dish as a JSON object to the user
    res.json(updatedDish);
  } catch (err) {
    res.status(422).json({ message: err.message });
  }
});
```

#### Delete

The CRUD operation Delete corresponds to the HTTP method DELETE. It is used to remove a resource from the system.
Let’s say that the world avocado shortage has reached a critical point, and we can no longer afford to serve this modern delicacy at all. We should go into the database and delete the item that corresponds to **Avocado Toast**, which we know has an `id` of 1223.

Request:

`DELETE http://www.myrestaurant.com/dishes/:id`

The express route for this deleting request can be

```js
router.delete("/dishes/:id", async (req, res) => {
  const { id } = req.params;
  try {
    const dish = await PostModel.findByIdAndDelete(id);
    // the status code is 204 because we are not returning data or entities
    return res.status(204).json({ success: true });
  } catch (err) {
    res.status(422).json({ message: err.message });
  }
});
```

> ## Resources

Reviewer 2: For consistency, rename this to "References" and add a horizontal rule above it (markdown syntax is ---) as mentioned in contributing guidelines.

- [What is REST](https://restfulapi.net/)
- [See this](https://www.educative.io/blog/crud-operations#what) for more information on how CRUD operations are performed directly on `SQL` database.

> - [database schemas](https://www.educative.io/blog/what-are-database-schemas-examples#types)

Reviewer 2: Ideally these links should be those you have referred or copy/pasted from for the reading and are not required to be opened and ready by students. If you want to recommend any of these as a further reading, then mention them where relevant in the reading content as an inline link.

> ## Intoduction to database schemas

Reviewer 3: in general titles should be: Cased Like This, but connecting works like `and, to`, etc are not capitalized

> ## Intoduction to database schemas

Reviewer 3: typo in introduction

run the whole thing through grammarly please

A database schema is a blueprint or architecture of how our data will look. It doesn’t hold data itself, but instead describes the shape of the data and how it might relate to other tables or models.

For example, this is a `customer` schema in MongoDB using mongoose.

```js
const mongoose = require("mongoose");

const customer = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  zipcode: {
    type: Number,
  },
});

module.exports = mongoose.model("Customer", customer);
```

The same `customer` schema in a MySql database.

```sql
CREATE TABLE customer (
 id INT AUTO_INCREMENT PRIMARY KEY,
 name VARCHAR(50) NOT NULL,
 zipcode INT(15) default NULL,
)

```

> We can see that the schema definition for the `customer` has a `name` which is a type `string` and a `zipcode` that is of type `number`.

Reviewer 3: ```suggestion
We can see that the schema definition for the `customer` has a `name` that is type `string` and a `zipcode` that is  type `number`.
```

## Introduction to CRUD operations

> When we are building APIs, we want our models to provide four basic types of functionality. The model must be able to Create, Read, Update, and Delete resources. Computer scientists often refer to these functions by the acronym `CRUD`. A model should have the ability to perform at most these four functions in order to be complete.

Reviewer 3: CRUD is not code so don't use `CRUD`, bold it or something.

> ### CRUD and REST

Reviewer 3: you have two headers of the same level (h3) next to each other

Reviewer 3: here and line 42

### What is REST?

> REST is a set of architectural constraints. API developers can implement REST in a variety of ways. When a client request is made via a RESTful API, it transfers a representation of the state of the resource to the requester or endpoint. This information, or representation, is delivered in one of several formats, JSON is the most generally popular file format to use.

Reviewer 3: don't quite like this phrase "transfers a representation of the state of the resource", not sure i even understand it :) i mean i understand what it means but i think it will be unnecessarily complex for students, perhaps find a simpler way to explain

### CRUD (CREATE, READ, UPDATE, DELETE)

> In a REST environment, CRUD often corresponds to the HTTP methods `POST`, `GET`, `PUT`, and `DELETE`, respectively. These are the fundamental elements of a persistent storage system.

Reviewer 3: are HTTP methods explained in a previous unit? if not, then people will be confused what an HTTP method means too

For example, imagine we are working with a system that is keeping track of meals and their corresponding prices for a restaurant. Let’s look at how we would implement CRUD operations.

#### Create

To create resources in a REST environment, we most commonly use the HTTP POST method. POST creates a new resource of the specified resource type.

> Imagine that we are adding a new food item to the stored list of dishes for this restaurant, and the dish objects are stored in a dishes resource.

Reviewer 3: you should define what "resource" means somewhere

Reviewer 2: Done in module 1

Note: Use `POST` when you want to add a child resource under resources collection.

##### Request:

`POST http://www.myrestaurant.com/api/dishes/`

We will need to send the `dish` data too.

#### Read

> To read resources in a REST environment, we use the GET method. in practice, reading a resource shouldn't change any information - it should only retrieve it.

Reviewer 3: ```suggestion
To read resources in a REST environment, we use the GET method. In practice, reading a resource shouldn't change any information - it should only retrieve it.
```

##### Request:

`GET http://www.myrestaurant.com/api/dishes/`

#### Update

PUT is the HTTP method used for the CRUD operation, Update.
So if the price of Avocado Toast has gone up, we should go into the database and update that information using PUT.

Note: Use `PUT` when you want to modify a singular resource which is already a part of resources collection. `PUT` replaces the resource with the data you send in its entirety.

##### Request:

`PUT http://www.myrestaurant.com/dishes/:id`

#### Delete

The CRUD operation Delete corresponds to the HTTP method DELETE. It is used to remove a resource from the system.
Let’s say that the world avocado shortage has reached a critical point, and we can no longer afford to serve this modern delicacy at all. We should go into the database and delete the item that corresponds to `Avocado Toast`, which we know has an `id` of 1223.

##### Request:

> `DELETE http://www.myrestaurant.com/dishes/:id`

Reviewer 3: if you are doing this on one line by itself you probably want triple backticks, here and everywhere else

## Introduction to CRUD operations

When we are building APIs, we want our models to provide four basic types of functionality. The model must be able to Create, Read, Update, and Delete resources. Computer scientists often refer to these functions by the acronym `CRUD`. A model should have the ability to perform at most these four functions in order to be complete.

### CRUD and REST

### What is REST?

REST is a set of architectural constraints. API developers can implement REST in a variety of ways. When a client request is made via a RESTful API, it transfers a representation of the state of the resource to the requester or endpoint. This information, or representation, is delivered in one of several formats, JSON is the most generally popular file format to use.

### CRUD (CREATE, READ, UPDATE, DELETE)

In a REST environment, CRUD often corresponds to the HTTP methods `POST`, `GET`, `PUT`, and `DELETE`, respectively. These are the fundamental elements of a persistent storage system.

For example, imagine we are working with a system that is keeping track of meals and their corresponding prices for a restaurant. Let’s look at how we would implement CRUD operations.

> - #### Create

Reviewer 3: i think this was fine before, i dont think using a bullet point followed by a header is a common practice

To create resources in a REST environment, we most commonly use the HTTP POST method. POST creates a new resource of the specified resource type.
Imagine that we are adding a new food item to the stored list of dishes for this restaurant, and the dish objects are stored in a dishes resource.

Note: Use `POST` when you want to add a child resource under resources collection.

Request:
`POST http://www.myrestaurant.com/api/dishes/`

We will need to send the `dish` data too.
`

- #### Read

To read resources in a REST environment, we use the GET method. in practice, reading a resource shouldn't change any information - it should only retrieve it.

> Note: technically, you can change data in a `GET` request but since we are creating a RESTful API, you shouldn't do that, and, in-general, changing data in a `GET` request would be confusing to the users of your api.

Reviewer 3: ```suggestion
To read resources in a REST environment, we use the GET method. Reading a resource should never change any information - it should only retrieve it. REST itself is more like a set of guidelines. Technically, you can change data in a `GET` request, but since we are creating a RESTful API, you shouldn't do that. Having a GET request that updates data in your database would be confusing to the users of your API and violate the expected behavior of REST.
```

Request:
`GET http://www.myrestaurant.com/api/dishes/`

> - #### Update

Reviewer 3: fix
